llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,847 @@
+      break;
+
+    case Rule::RKind::K_SortedIntegersListMetadata:
+      MetadataName = R.getStorage<Rule::RKind::K_IntegersListMetadata>();
+      if (F->hasMetadata(MetadataName)) {
+        MDNode *MDN = F->getMetadata(MetadataName);
+
+        SmallVector<std::uint64_t, 8> Values;
+        for (const MDOperand &MDOp : MDN->operands())
+          Values.push_back(mdconst::extract<ConstantInt>(MDOp)->getZExtValue());
+
+        llvm::sort(Values);
+
+        for (std::uint64_t V : Values)
+          Result += "-" + std::to_string(V);
+      }
+      break;
+
+    case Rule::RKind::K_FlagAttribute:
+      Data = R.getStorage<Rule::RKind::K_FlagAttribute>();
+      if (F->hasFnAttribute(Data.Name))
+        Result += Data.IfPresentStr;
+      else
+        Result += Data.IfAbsentStr;
+      break;
+    }
+
+    Result += "-";
+  }
+
+  return static_cast<std::string>(Result);
+}
+} // namespace
+
+std::unique_ptr<ModuleSplitterBase>
+getDeviceCodeSplitter(ModuleDesc MD, IRSplitMode Mode,
+                      bool EmitOnlyKernelsAsEntryPoints) {
+  FunctionsCategorizer Categorizer;
+
+  EntryPointsGroupScope Scope = selectDeviceCodeGroupScope(Mode);
+
+  switch (Scope) {
+  case Scope_Global:
+    // We simply perform entry points filtering, but group all of them together.
+    Categorizer.registerRule(
+        [](Function *) -> std::string { return GLOBAL_SCOPE_NAME; });
+    break;
+  case Scope_PerKernel:
+    // Per-kernel split is quite simple: every kernel goes into a separate
+    // module and that's it, no other rules required.
+    Categorizer.registerRule(
+        [](Function *F) -> std::string { return F->getName().str(); });
+    break;
+  case Scope_PerModule:
+    // The most complex case, because we should account for many other features
+    // like aspects used in a kernel, large-grf mode, reqd-work-group-size, etc.
+
+    // This is core of per-source device code split
+    Categorizer.registerSimpleStringAttributeRule(ATTR_SYCL_MODULE_ID);
+
+    // This attribute marks virtual functions and effectively dictates how they
+    // should be groupped together. By design we won't split those groups of
+    // virtual functions further even if functions from the same group use
+    // different optional features and therefore this rule is put here.
+    // Strictly speaking, we don't even care about module-id splitting for
+    // those, but to avoid that we need to refactor the whole categorizer.
+    // However, this is good enough as it is for an initial version.
+    // TODO: for AOT use case we shouldn't be outlining those and instead should
+    // only select those functions which are compatible with the target device
+    Categorizer.registerSimpleStringAttributeRule("indirectly-callable");
+
+    // Optional features
+    // Note: Add more rules at the end of the list to avoid chaning orders of
+    // output files in existing tests.
+    Categorizer.registerSimpleStringAttributeRule("sycl-register-alloc-mode");
```

Comment:
Looking at this code once more, I believe that it is over-engineered. It was written (by me) in assumption that there will be more different code splitters as we add new features, but as it turned out, any nested splitting is a huge pain to deal with. Therefore, we have those hundreds lines of code which could be simplified down to a single function which won't even be that long even without all those helpers about registering rules which we have.

Moreover, there are cases where we could just stop and drop the rest of the rules if we see certain attributes. See the comment around `indirectly-callable` handling above.

Therefore, I definitely would not object some refactoring in this area to keep the code more straightforward.

Reply:
Removed FunctionCategorizer.